src/e.uv-link-t.md

```diff
@@ -0,0 +1,228 @@
+---
+title: uv_link_t - libuv pipeline
+date: 2016-08-15
+---
+
+## Preface
+
+Writing servers/clients in C could be non-trivial. Even with the help of such
+powerful (and awesome dinosaur) libraries as [libuv][0], it still takes lots of
+effort and boilerplate code to create real world applications.
+
+Some of this boilerplate code comes from the use of the widespread protocols
+like TLS (SSL) and HTTP. While there are popular implementations available
+as an Open Source libraries ([OpenSSL][1], [http-parser][2]), they still either
+provide very abstract interface (like [http-parser][2]), or an API to transfer
+the responsibility of the networking to the library itself (like `SSL_set_fd()`
+in [OpenSSL][1] and Amazon's [s2n][3]). Such abstract nature makes them easier
+to embed, but the adaptor code inevitably tend to appear in the particular
+applications.
+
+## Precursor - StreamBase
+
+[libuv][0] is hardly an exception, and [node.js][4] and [bud][5]'s TLS
+implementation is a vivid evidence of this. However, in a contrast to [bud][5],
+[node.js][4] TLS code lives off on an abstraction called [StreamBase][6]. By
+separating [libuv][0]-specific adaptor code into a generic C++ class, we have
+created a foundation for a simpler and reusable implementation of any other
+protocol! See, for example, recent [node_http_parser.cc][7] which uses only
+a minor amount of power available through the means of [StreamBase][6], but
+nevertheless provides [10-20%][8] performance improvement since its inception.
+
+This implementation has some major drawbacks, preventing its wider adoption
+outside of the node.js core:
+
+* C++ headers: lots of virtual classes, complex API, non-trivial inheritance
+  scheme
+* High internal dependence on the node.js core itself
+
+Because of these issues (and my own limitations) [StreamBase][6] has defied all
+attempts to make it public.
+
+## uv_link_t
+
+Heavily inspired by the success of [StreamBase][6] in the node.js core, a
+[uv_link_t][9] library was created. It has lots of similarities with the
+[StreamBase][6], but it is:
+
+* Implemented in C: self-documented structures, C-cast based inheritance, etc
+* Standalone library
+
+The API is based on the [uv_stream_t][10] and shouldn't come as a big surprise
+to the users, since [uv_link_t][9] is intended to be used together with
+[libuv][0].
+
+Here is a visual explanation of how [uv_link_t][0] works:
+
+![uv_link_source_t][17]
+
+## Examples
+
+Before we will take a peek look at the APIs, let's discuss what can be done with
+[uv_link_t][9]. Technically, any stream-based (TCP-based) protocol can be
```

Comment:
would it be accurate to say "any stream-based (i.e. anything that uses `uv_stream_t`)"?